src/Cli/dotnet/Commands/Run/RunCommand.cs

```diff
@@ -346,24 +383,31 @@ private MSBuildArgs SetupSilentBuildArgs(MSBuildArgs msbuildArgs)
         return msbuildArgs;
     }
 
-    internal ICommand GetTargetCommand(Func<ProjectCollection, ProjectInstance>? projectFactory)
+    /// <summary>
+    /// Gets the target command and project instance for the specified project factory.
+    /// The Project is used to fill in any 'holes' in the arguments from the launch profile chosen, if any.
+    /// </summary>
+    /// <param name="projectFactory"></param>
+    /// <returns></returns>
+    /// <exception cref="GracefulException"></exception>
+    internal (Utils.Command command, ProjectInstance? project) GetTargetCommand(Func<ProjectCollection, ProjectInstance>? projectFactory)
     {
         if (projectFactory is null && ProjectFileFullPath is null)
         {
             // If we are running a file-based app and projectFactory is null, it means csc was used instead of full msbuild.
             // So we can skip project evaluation to continue the optimized path.
             Debug.Assert(EntryPointFileFullPath is not null);
-            return CreateCommandForCscBuiltProgram(EntryPointFileFullPath);
+            return (CreateCommandForCscBuiltProgram(EntryPointFileFullPath), null);
```

Comment:
> It's worth noting that the direct-Csc usage of dotnet run doesn't support this feature because there is no ProjectInstance. It's possible that this gap could be filled with a proxy or a static set of Properties or something.

Another option would be to fall back to MSBuild if there are any `$(...)` holes in the launch profile to ensure correctness.

I think we should file a follow up issue to track this.